server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        rsync_args = self.make_rsync_args(tmp_path1, str(self.src_prefix),
+                                          self.dest_prefix, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
+        _logger.info(this_output)
+        if not is_successful:
+            _logger.error(this_output)
+            return (is_successful, this_output)
+        output += this_output
+        return (is_successful, output)
+
+    def process_main(self):
+        (successful, output) = self.rsync()
+        if not successful:
+            raise PulpCodedException(message=output)
+
+
+class Publisher(PublishStep):
+    """
+    RSync publisher class that provides the common code for publishing to remote server. Each
+    plugin needs to extend this class and add necesary logic to the _add_necesary_steps method.
+    """
+    def __init__(self, repo, publish_conduit, config, distributor_type):
+        """
+        :param repo: Pulp managed Yum repository
+        :type  repo: pulp.plugins.model.Repository
+        :param publish_conduit: Conduit providing access to relative Pulp functionality
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config: Pulp configuration for the distributor
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+
+        super(Publisher, self).__init__("Repository publish", repo,
+                                        publish_conduit, config,
+                                        distributor_type=distributor_type,
+                                        working_dir="")
+
+        distributor = Distributor.objects.get_or_404(repo_id=self.repo.id,
+                                                     distributor_id=publish_conduit.distributor_id)
+        self.last_published = distributor["last_publish"]
+        self.last_deleted = repo.last_unit_removed
+        self.repo_type = repo.notes["_repo-type"]
+        self.predistributor = self._get_predistributor()
+        if self.last_published and not isinstance(self.last_published, basestring):
+            string_date = dateutils.format_iso8601_datetime(self.last_published)
+        elif self.last_published:
+            string_date = self.last_published
+        else:
+            string_date = None
+        if self.predistributor:
+            search_params = {'repo_id': repo.id,
+                             'distributor_id': self.predistributor["id"],
+                             'started': {"$gte": string_date}}
+            self.predist_history = RepoPublishResult.get_collection().find(search_params)
+        else:
+            self.predist_history = []
+
+        self.remote_path = self.get_remote_repo_path()
+
+        if not self.is_fastforward():
+            date_filter = None
+        else:
+            range_criteria = {START_DATE_KEYWORD: self.last_published,
+                              END_DATE_KEYWORD: self.predistributor["last_publish"]}
```

Comment:
Since ``self.predistributor`` is (I think) a proper object, I think we'd prefer accessing with normal attribute notation. Minor suggestion, but it makes it more clear what we're dealing with.